### ISSUE TITLE
orbit() can be transcendentally cheaper than it currently is for client rendering

### ISSUE THREAD

**Issue by Incoming5643** [tgstation/-tg-station#13697]
<s>http://www.byond.com/forum/?post=1998783</s>

The system now: Call animate every() 0.2 seconds (forever) to move an atom along a path that approximates a circle.

The <s>lummox</s> solution: Translate the icon of an atom by the radius of the orbit, then just spin the icon with a single SpinAnimation call to get a literal circle.

For the love of god someone who knows matrices, PLEASE

**Comment by tkdrg** [comment]
you don't even need to know matrices for this, just use SpinAnimation()

**Comment by RemieRichards** [comment]
Lummox's suggestion is to make the parent spin too.
This is not at all what we want.

**Comment by Incoming5643** [comment]
Well I mean you don't have to rotate the parent object. You could just rotate the atom over the stationary parent. The important idea here is that a rotating atom visually offset from the parent that uses a rotate animation would still appear to be an orbit.

**Comment by RemieRichards** [comment]
That's not what Lummox is saying:

>Why not make an overlay that's translated out a certain distance, and use a rotating animation on the parent object?

He's saying to make an OVERLAY, on the parent object, you can't (yes, CAN'T) rotate that without rotating the parent OR awkwardly juggling it out of the parent's overlay list every tick (or smooth value like the 0.2 used by ```orbit()```)

Not to mention, to be able to rotate it, it would need to be an ```/icon```, and ```/icon```s aren't cheap themselves, you'd be shifting the cost from 1 user's computer to the server itself, potentially slowing everything, which on it's own, no, isn't that bad, but as you complain about orbit being overused, the same logic applies here.

**Comment by Incoming5643** [comment]
Alright well lets, like, just forget what lummox said.

Lock the atom's loc to the parent's, as is already done
Translate() the atom's matrix off to the side
Then have the atom use SpinAnimation() (three animate() calls total) to rotate.

It's all matrices and animate(). Would that not work?

**Comment by RemieRichards** [comment]
That should work, yes.

**Comment by Incoming5643** [comment]
okay then I'll try it when it's not 5 am unless someone beats me to it!

Good night :full_moon_with_face:

**Comment by Incoming5643** [comment]
Also to be clear I'm not looking to scrub out that version of orbit either, there's still fancier things that it can do that this implementation probably wouldn't be great at. It's just for bog standard "fixed radius with an indefinite lifespan" I think we could cut down the costs.

Adding things like particle effects to wizard spells would still be really cool.

**Comment by RemieRichards** [comment]
@Incoming5643 I see no situation ```neworbit()``` can't handle that ```orbit()``` can.
By default, the rotation will look different, but if you wanted current ghost orbit for example, you'd just rotate the ghost's transform by 90 (or -90) and THEN call SpinAnim.

tl;dr ```neworbit()``` should be entirely superior and I'll write it up with some args when I get home (2 hours or so) (no it wont be called ```neworbit()```, ```orbit()```'s insides will be changed)

**Comment by RemieRichards** [comment]
I did discover one thing it can't do, specific increments, but I'm not entirely sure they worked, and they probably ruined the rotation effect anyway.(we only used the basic 15 increment)

**Comment by RemieRichards** [comment]
Ok so ```SpinAnimation()``` doesn't work...

![triangles!?](http://i.imgur.com/NzuT9UR.gifv)

it *IS* a neat effect though...

**Comment by RemieRichards** [comment]
![HEXAGON!?!](http://i.imgur.com/6gQ9fvJ.gif)

This one's a hexagon!
(Modified ```SpinAnimation()``` with 6 matrices + animate calls rather than 3)

I can *try* to find the best approximate for a circle, but a true circle would be 360 animate calls! (Which still might be less than what's already done by ```orbit()```)

One idea might be a supplyable proc argument, allowing us to (legitimately) obtain these triangular and hexagonal rotations though.

**Comment by RemieRichards** [comment]
@Incoming5643 I've produced something that allows for reasonable circular orbits (and can easily produce the other shapes!!) that runs with the same CPU amount, and 36 Matrix calls (in this case the orbit is 36 segments) vs. infinite matrix calls, logically this would be faster, but I'd like to test it on you!

**Comment by RemieRichards** [comment]
Ok so I tested it with a proc in the middle (MSO's suggestion):
for current orbit() 10 rotations around someone is 240 odd animate() calls

for new orbit() at 23 rotations (tests based on time, not rotation amounts) it was 36, which is the number of segments used, meaning that's a -constant- 35 calls and then never any more, obviously this stacks with ghosts but it should be MUCH nicer.

**Comment by Incoming5643** [comment]
Anything that that replaces continuous animate calls with a finite number should solve the problem. Even my potato can deal with a large number of animate calls over a short period, it's a sustained load that starts to choke it up. I don't think it matters too much if the number of animate call up front is larger. 35 calls is nothing when you consider that the old method would use more after just 7 seconds of looking at it.

**Comment by RemieRichards** [comment]
@Incoming5643 ye, I'm making the PR soon (with ~features~) so be sure to shill!
